File: batou-rules/testdata/fixtures/python/vulnerable/ssrf_basic.py
# Vulnerable: SSRF via requests.get with user-controlled URL
# Expected: Taint sink match for py.requests.get (CWE-918)

import requests
from flask import Flask, request, jsonify

app = Flask(__name__)


@app.route("/fetch")
def fetch_url():
    url = request.args.get("url")
    response = requests.get(url)
    return response.text


@app.route("/proxy")
def proxy():
    target = request.args.get("target")
    resp = requests.get(target, timeout=10)
    return jsonify({"status": resp.status_code, "body": resp.text[:1000]})


@app.route("/check-health")
def health_check():
    service_url = request.form.get("service_url")
    try:
        resp = requests.get(service_url, timeout=5)
        return jsonify({"healthy": resp.status_code == 200})
    except Exception as e:
        return jsonify({"healthy": False, "error": str(e)})


@app.route("/webhook-test", methods=["POST"])
def test_webhook():
    webhook_url = request.json.get("url")
    payload = request.json.get("payload", {})
    resp = requests.post(webhook_url, json=payload)
    return jsonify({"status": resp.status_code})

File: batou-rules/testdata/fixtures/javascript/vulnerable/ssrf_basic.ts
import { Request, Response } from 'express';
import axios from 'axios';
import fetch from 'node-fetch';

// VULNERABLE: SSRF via user-controlled URL passed to axios
export async function proxyRequest(req: Request, res: Response): Promise<void> {
  const targetUrl = req.query.url as string;

  // VULNERABLE: axios.get with user-controlled URL (SSRF)
  const response = await axios.get(targetUrl);
  res.json({ data: response.data });
}

// VULNERABLE: SSRF via fetch with user-controlled URL
export async function fetchWebhook(req: Request, res: Response): Promise<void> {
  const webhookUrl = req.body.callback_url;

  // VULNERABLE: fetch with user-controlled URL
  const response = await fetch(webhookUrl, {
    method: 'POST',
    body: JSON.stringify({ status: 'complete' }),
    headers: { 'Content-Type': 'application/json' },
  });

  res.json({ status: response.status });
}

// VULNERABLE: SSRF with redirect following enabled
export async function fetchPage(req: Request, res: Response): Promise<void> {
  const pageUrl = req.query.url as string;

  const response = await axios.get(pageUrl, {
    maxRedirects: 10,
    followRedirects: true,
  });

  res.send(response.data);
}
